### PR TITLE
Email reports enhancement

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | n
| Related developer documentation PR URL | n
| Issues addressed (#s or URLs) | #990
| BC breaks? | n
| Deprecations? | n

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
PR added Send count, Read ratio, Clicks, Unique clicks, Clicks ratio, Unique clicks ratio, Unsubscribed, Unsubscribed ratio to email reports.

#### Steps to test this PR:
1. Create Email report and select from columns new options   Send count, Read ratio, Clicks, Unique clicks, Clicks ratio, Unique clicks ratio, Unsubscribed, Unsubscribed ratio
2. See report

#990

### CODE REVIEW

**Comment by luizeof** [comment]
This new report fields are available on Graphs ?

**Comment by kuzmany** [comment]
@luizeof  table.

**Comment by dongilbert** [comment]
This works well, but I was required to make some changes in order to get it working. See inline comments.

File: app/bundles/EmailBundle/EventListener/ReportSubscriber.php [hunk 1]
```diff
@@ -159,11 +224,20 @@ public function onReportGenerate(ReportGeneratorEvent $event)
...
+            ->groupBy('cut2.channel_id');
 
         switch ($context) {
             case 'emails':
                 $qb->from(MAUTIC_TABLE_PREFIX.'emails', 'e')
-                    ->leftJoin('e', MAUTIC_TABLE_PREFIX.'emails', 'vp', 'vp.id = e.variant_parent_id');
+                    ->leftJoin('e', MAUTIC_TABLE_PREFIX.'emails', 'vp', 'vp.id = e.variant_parent_id')
+                    ->leftJoin('e', sprintf('(%s)', $qbcut->getSQL()), 'cut', 'e.id = cut.channel_id')
+                    ->leftJoin('e', MAUTIC_TABLE_PREFIX.'lead_donotcontact', 'dnc', 'e.id = dnc.channel_id and dnc.channel=\'email\' and dnc.reason=1')
+                    ->groupBy('e.id');
```
Comment (dongilbert) [inline app/bundles/EmailBundle/EventListener/ReportSubscriber.php]: The query on line 246 required having the following lines added in order to function:

```
                  ->leftJoin('e', sprintf('(%s)', $qbcut->getSQL()), 'cut', 'e.id = cut.channel_id')
                    ->leftJoin('e', MAUTIC_TABLE_PREFIX.'lead_donotcontact', 'dnc', 'e.id = dnc.channel_id and dnc.channel=\'email\' and dnc.reason=1')
                    ->groupBy('e.id, es.date_sent, es.date_read, es.email_address');
```

**Comment by dongilbert** [comment]
I pushed the requested changes to your branch. Can you test?

**Comment by alanhartless** [comment]
While testing this, I found some issues with the new columns/queries when it came to individual email stats. For example, the hits and unique hits were email wide rather than the "clicks" for the specific contact which is what I think users will expect with this feature. The click ratios also didn't account for multiple trackable URLs per email so the percent could be greater than 100% if more than one link was clicked per email. Finally, the unsubscribed column would filter out any record without a do not contact record. These were fixed in the last commit.

**Comment by escopecz** [comment]
Works good for me :+1:

**Comment by MxyzptlkFishStix** [comment]
Is the open rate using this formula?

(Opens + Clicks w/o Opens) / (Total Emails Sent - Bounces)

**Comment by kuzmany** [comment]
Nope. Bounces are stored just for email address.

**Pull request merged**